### PR TITLE
Adding ability to support all different code coverage formats with V2

### PR DESCRIPTION
**Pull request by vinayakmsft**
**PR description -** Here, we are adding the abilities to support all the code coverage formats with the PCCRV2. So with these changes, we can now pass the .coverage/.covx files as well to the PCCRV2 and those would be converted to the cjson format and we would get the cjson file view in the code coverage tab.

Conversion logic implemented - Inorder to convert .coverage /.covx file to .cjson , we will have to first convert them into the .xml format and then we would be able to convert it into the .cjson format and upload to the logstore. 

**Before:**

Earlier when we used to pass the .coverage,.covx files as input to the PCCRV2 task, the PCCRV2 task used to fail as it was not able to parse these coverage file formats. These formats were not supported in the PCCRV2 task. 

**After:** With these new changes, now the PCCRV2 task can successfully handle these different coverage file formats and then upload a single common file (.cjson ) to the logstore. 

**CC tab for .coverage file passed as input to the PCCRV2 task**

image

**CC tab for .covx file passed as input to PCCRV2**

image

### CODE REVIEW

File: src/CoveragePublisher/Parsers/CoverageParserTools/PublisherCoverageFileConfiguration.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
+﻿﻿using Microsoft.Azure.Pipelines.CoveragePublisher.Model;
+using Microsoft.CodeCoverage.Core;
+using Microsoft.CodeCoverage.IO.Coverage;
+using Microsoft.TeamFoundation.TestManagement.WebApi;
+using Palmmedia.ReportGenerator.Core;
```
Comment (eswarnathprudhivi) [inline src/CoveragePublisher/Parsers/CoverageParserTools/PublisherCoverageFileConfiguration.cs:5]: Do  we need report generator imports here?

File: src/CoveragePublisher/Parsers/CoverageParserTools/PublisherCoverageFileConfiguration.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        public bool FixCoverageBuffersMismatch { get; set; }
+
+        /// <inheritdoc/>
+        public int MaxDegreeOfParallelism { get; set; } = 10;
+
+        public bool SkipInvalidData { get; set; }
+
+        public CoverageMergeOperation MergeOperation { get; set; }
+
+        internal static PublisherCoverageFileConfiguration Default { get; } = new PublisherCoverageFileConfiguration
```
Comment (eswarnathprudhivi) [inline src/CoveragePublisher/Parsers/CoverageParserTools/PublisherCoverageFileConfiguration.cs:44]: This is not the right way, don't do this way, try to create an object from helper class

File: src/CoveragePublisher/Parsers/CoverageParserTools/PublisherCoverageFileConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        /// <inheritdoc/>
+        public bool ReadSkippedFunctions { get; set; }
+
+        /// <inheritdoc/>
+        public bool ReadSnapshotsData { get; set; }
+
+        /// <inheritdoc/>
+        public bool GenerateCoverageBufferFiles { get; set; }
+
+        /// <inheritdoc/>
```
Comment (eswarnathprudhivi) [inline src/CoveragePublisher/Parsers/CoverageParserTools/PublisherCoverageFileConfiguration.cs:34]: Please remove inheritdoc comments

File: src/CoveragePublisher/Parsers/CoverageParserTools/PublisherCoverageFileConfiguration.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+using Palmmedia.ReportGenerator.Core.Parser.Filtering;
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Diagnostics;
+using System.IO;
+
+namespace Microsoft.Azure.Pipelines.CoveragePublisher.Parsers.CoverageParserTools
+{
+    internal class PublisherCoverageFileConfiguration : ICoverageFileConfiguration
```
Comment (eswarnathprudhivi) [inline src/CoveragePublisher/Parsers/CoverageParserTools/PublisherCoverageFileConfiguration.cs:17]: shouldn't be internal class

File: src/CoveragePublisher/Parsers/CoverageParserTools/ReportGeneratorTool.cs [hunk 5]
```diff
@@ -70,6 +81,72 @@ public List<FileCoverageInfo> GetFileCoverageInfos()
...
+                    }
+                }
+            }
+
+            return fileCoverages;
+        }
+
+      private async Task<List<string>> TransformCoverageFilesToXml(IList<string> inputCoverageFiles, CancellationToken cancellationToken)
+        {
+            // Customers like intune invoke vstest.console.exe multiple times inside a single job. Transform cov files resulting from each run into a different subdirectory
```
Comment (eswarnathprudhivi) [inline src/CoveragePublisher/Parsers/CoverageParserTools/ReportGeneratorTool.cs:127]: Is this comment required?

File: src/CoveragePublisher/Parsers/CoverageParserTools/ReportGeneratorTool.cs [hunk 6]
```diff
@@ -70,6 +81,72 @@ public List<FileCoverageInfo> GetFileCoverageInfos()
...
+            var utility = new CoverageFileUtilityV2(PublisherCoverageFileConfiguration.Default);
```
Comment (eswarnathprudhivi) [inline src/CoveragePublisher/Parsers/CoverageParserTools/ReportGeneratorTool.cs:128]: change the way you are picking the PublisherCoverageFileConfiguration

File: src/CoveragePublisher/Parsers/CoverageParserTools/ReportGeneratorTool.cs [hunk 7]
```diff
@@ -70,6 +81,72 @@ public List<FileCoverageInfo> GetFileCoverageInfos()
...
+
+            var transformedXmls = new List<string>();
+            foreach (var nativeCoverageFile in inputCoverageFiles)
```
Comment (eswarnathprudhivi) [inline src/CoveragePublisher/Parsers/CoverageParserTools/ReportGeneratorTool.cs:131]: any reason it is named as nativeCoverageFile? can you name it coverageFile

File: src/CoveragePublisher/Parsers/CoverageParserTools/ReportGeneratorTool.cs [hunk 8]
```diff
@@ -70,6 +81,72 @@ public List<FileCoverageInfo> GetFileCoverageInfos()
...
+            List<FileCoverageInfo> fileCoverages = new List<FileCoverageInfo>();
+
+            if (Configuration.CoverageFiles == null)
+            {
+                return fileCoverages;
+            }
+
+            var transformedXml = TransformCoverageFilesToXml(Configuration.CoverageFiles, token);
+
+            _parserResult = ParseCoverageFiles(transformedXml.Result);
```
Comment (eswarnathprudhivi) [inline src/CoveragePublisher/Parsers/CoverageParserTools/ReportGeneratorTool.cs:97]: what does Result has?, transformedXml is a list of string right?

File: src/CoveragePublisher/Parsers/CoverageParserTools/PublisherCoverageFileConfiguration.cs (hunk 1 above)
Comment (vinayakmsft) [reply src/CoveragePublisher/Parsers/CoverageParserTools/PublisherCoverageFileConfiguration.cs]: Removed

File: src/CoveragePublisher/Parsers/CoverageParserTools/ReportGeneratorTool.cs [hunk 9]
```diff
@@ -23,6 +23,7 @@
 using CoverageStatus = Microsoft.TeamFoundation.TestManagement.WebApi.CoverageStatus;
 using Microsoft.CodeCoverage.IO.Coverage;
 using Microsoft.CodeCoverage.IO;
+using SixLabors.Fonts.Tables.TrueType;
```
Comment (eswarnathprudhivi) [inline src/CoveragePublisher/Parsers/CoverageParserTools/ReportGeneratorTool.cs:26]: please check this and remove this

File: src/CoveragePublisher/Parsers/CoverageParserTools/ReportGeneratorTool.cs [hunk 10]
```diff
@@ -122,22 +123,25 @@ public List<FileCoverageInfo> GetFileCoverageInfos(CancellationToken token)
             return fileCoverages;
         }
 
+
       private async Task<List<string>> TransformCoverageFilesToXml(IList<string> inputCoverageFiles, CancellationToken cancellationToken)
         {
-            // Customers like intune invoke vstest.console.exe multiple times inside a single job. Transform cov files resulting from each run into a different subdirectory
-            var utility = new CoverageFileUtilityV2(PublisherCoverageFileConfiguration.Default);
+            
+            PublisherCoverageFileConfiguration GenericCoverageFileConfiguration = new PublisherCoverageFileConfiguration(true,true,true,true,true,true);
```
Comment (eswarnathprudhivi) [inline src/CoveragePublisher/Parsers/CoverageParserTools/ReportGeneratorTool.cs:130]: Please call it this way new PublisherCoverageFileConfiguratoin( ReadModules : true, ReadSkippedModules: true)

**Review by eswarnathprudhivi — APPROVED** [review]
lgtm

**Pull request merged**